### DISCUSSION TITLE
Order of variables interpolation in services during deploy

### DISCUSSION THREAD

**Discussion by danibix95** [mia-platform/community#72] [Q&A] [answered]
Hi,

I encountered an issue with the order of variables interpolation while deploying a set of microservices. 
In particular, I defined a variable called `LOG_LEVEL` in the `public variables` section as shown below:

![Screenshot from 2022-11-07 14-25-52](https://user-images.githubusercontent.com/3228581/200325095-b35be59a-2b29-4415-bc23-232b9454c011.png)

As it can be seen, its value is `info` in all the available environments.

Afterwards I deployed my project's services in the `development` environment to conduct some tests. I started observing services logs in the runtime section to monitor their responses and I noticed that a great amount of records where produced, especially at `debug` level, which is lower than the level configured in the public variables. Consequently, no `debug` log should be produced. 

To further explore this issue, I opened the `runtime` section and I had a look at the `describe` tab of the pod, which reported the following config (to describe the issue I will use the `crud-service`, but all the deployed services were behaving in the same manner):

![Screenshot from 2022-11-07 14-57-28](https://user-images.githubusercontent.com/3228581/200328215-2091acb0-5e1a-4950-aeec-2674d272bcd6.png)

As it can be observed, the `LOG_LEVEL` environment variable is set to `trace` instead of the value `info` I previously defined in the `public variables` section.

Therefore, I reviewed my configuration to ensure all the services were correctly set up, that is they needed to adopt as value of environment variable `LOG_LEVEL` the one interpolated from _public_ variables. As shown in the following screenshot of the `crud-service` config, its configuration was fine. Similarly, all the other services of my project were configured in the same manner.

![Screenshot from 2022-11-07 14-47-07](https://user-images.githubusercontent.com/3228581/200327465-9adf3124-2385-4c0b-9620-37408895346a.png)

For this reason, I started suspecting that some `project setting` environmental variable was interfering with the public variables. Checking the environmental variables defined in the in that section led to no further clue, since no `LOG_LEVEL` variable was defined.

However, diving into the git provider configuration associated to my project, I observed that my project was inheriting a `LOG_LEVEL` variable from its parent group, of which I do not have any access.  
**Could it be possible that such variable gets a higher priority than the ones defined in the project `public variables` settings?**

As explained in the [documentation](https://docs.mia-platform.eu/docs/development_suite/api-console/api-design/public_variables#inheritance-with-env-variables), _public_ variables should get a higher priority over the _secret_ ones defined in the `project settings` section.

As stated by the documentation current behaviour seems to be an interpolation bug. Is it possible to get a clarification on this topic?

**Comment by davidebianchi** [comment, marked as answer]
Hi @danibix95, it could be possible that the variable in one parent group is more specific?
If I correctly remember, the variable with the specific prefix of the environment will take precedence over the general variable.

So, for example:
- LOG_LEVEL at project level take precedence over LOG_LEVEL in parent level
- DEV_LOG_LEVEL at parent level take precedence over LOG_LEVEL at project level

The rule should work for both public and private variables. 

Can you try to use a specific env var?

**Reply by danibix95** [reply]
Yes, I confirm that the variable in the parent project is defined as `DEV_LOG_LEVEL`.  
As far as I know, this behaviour is not explicitly described in the documentation, so that I was confused about the interpolation outcome.

Following your recommendation, I defined a variable `DEV_LOG_LEVEL` in the public variables. I also left the services config unchanged, so that they obtain the value `LOG_LEVEL` from the interpolated variable. Effectively, now service's `LOG_LEVEL` environment variable is set to the correct value `info`.

However, I do find this configuration a little confusing, since `public variables` already provides a table with a column for each environment. Consequently I would expect that no prefix should be required to a variable to overwrite `project settings` variables.

What do you think?

**Reply by davidebianchi** [reply]
Yes, I agree, this could be confusing. From the technical point of view, once the env var is in the shell you don't know where are them from. So, it is not possible to handle with ease.

Another solution could be to add interpolation of the public env var directly from Console (or fetching it from API, instead of using the env var in the shell). In this way, we could remove this issue and better manage the variables. The interpolation of secret env would will be probably deprecated, because it is not possible to handle if you use another secret manager, for example Vault.

**Reply by danibix95** [reply]
OK, for the moment I'll keep to recommended change deploy with the correct value. Should we update the documentation to further clarify variables hierarchy?

**Reply by davidebianchi** [reply]
Yes, the documentation should be updated! Can you try to improve this part?
